modules/localization/msf/local_pyramid_map/base_map/base_map.cc

```diff
@@ -0,0 +1,584 @@
+  success = map_node_cache_lvl2_->Get(index, &node);
+  if (success) {
+    node->SetIsReserved(true);
+    map_node_cache_lvl1_->Put(index, node);
+    lock2.unlock();
+    return node;
+  }
+  lock2.unlock();
+
+  // load from disk
+  std::cerr << "GetMapNodeSafe: This node don't exist in cache! " << std::endl
+            << "load this node from disk now!" << index << std::endl;
+
+  LoadMapNodeThreadSafety(index, true);
+
+  boost::unique_lock<boost::recursive_mutex> lock3(map_load_mutex_);
+  map_node_cache_lvl2_->Get(index, &node);
+  map_node_cache_lvl1_->Put(index, node);
+  lock3.unlock();
+
+  return node;
+}
+
+bool BaseMap::IsMapNodeExist(const MapNodeIndex& index) {
+  boost::unique_lock<boost::recursive_mutex> lock(map_load_mutex_);
+  bool if_exist = map_node_cache_lvl1_->IsExist(index);
+  lock.unlock();
+  return if_exist;
+}
+
+bool BaseMap::SetMapFolderPath(const std::string folder_path) {
+  map_config_->map_folder_path_ = folder_path;
+  // Try to load the config
+  std::string config_path = map_config_->map_folder_path_ + "/config.xml";
+  if (cyber::common::PathExists(config_path)) {
+    map_config_->Load(config_path);
+    return true;
+  }
+  return false;
+}
+
+void BaseMap::AddDataset(const std::string dataset_path) {
+  map_config_->map_datasets_.push_back(dataset_path);
+  std::string config_path = map_config_->map_folder_path_ + "/config.xml";
+  map_config_->Save(config_path);
+}
+
+void BaseMap::LoadMapNodes(std::set<MapNodeIndex>* map_ids) {
+  if (map_ids->size() > map_node_cache_lvl1_->Capacity()) {
+    std::cerr << "map_ids's size is bigger than cache's capacity" << std::endl;
+    return;
+  }
+
+  // check in cacheL1
+  std::set<MapNodeIndex>::iterator itr = map_ids->begin();
+  while (itr != map_ids->end()) {
+    boost::unique_lock<boost::recursive_mutex> lock1(map_load_mutex_);
+    if (map_node_cache_lvl1_->IsExist(*itr)) {
+      // fresh LRU list
+      map_node_cache_lvl2_->IsExist(*itr);
+      lock1.unlock();
+      itr = map_ids->erase(itr);
+    } else {
+      lock1.unlock();
+      ++itr;
+    }
+  }
+
+  // check in cacheL2
+  itr = map_ids->begin();
+  BaseMapNode* node = nullptr;
+  while (itr != map_ids->end()) {
+    boost::unique_lock<boost::recursive_mutex> lock2(map_load_mutex_);
+    if (map_node_cache_lvl2_->Get(*itr, &node)) {
+      node->SetIsReserved(true);
+      map_node_cache_lvl1_->Put(*itr, node);
+      lock2.unlock();
+      itr = map_ids->erase(itr);
+    } else {
+      lock2.unlock();
+      ++itr;
+    }
+  }
+
+  // load from disk sync
+  std::vector<std::future<void>> load_futures_;
+  itr = map_ids->begin();
+  while (itr != map_ids->end()) {
+    AERROR << "Preload map node failed!";
+    load_futures_.emplace_back(
+        cyber::Async(&BaseMap::LoadMapNodeThreadSafety, this, *itr, true));
+    ++itr;
+  }
+
+  for (auto& future : load_futures_) {
+    if (future.valid()) {
+        future.get();
+    }
+  }
+
+  // check in cacheL2 again
```

Comment:
consider moving duplicate code to a common function (not critical)

Reply:
fixed